// views/BookmarkList.tsx
import {bookmarkActionsType} from 'actions/bookmarkActions';
import {observer} from 'mobx-react';
import {BookmarkModel} from 'models/BookmarkModel';
import {BookmarkViewModel} from 'models/BookmarkViewModel';
import * as React from 'react';
import {BookmarkStore} from 'stores/BookmarkStore';
import {Bookmark} from 'views/Bookmark';

export interface PropsType {
  bookmarkActions: bookmarkActionsType;
  bookmarkStore: BookmarkStore;
  bookmarkView: BookmarkViewModel;
}

@observer
export class BookmarkList extends React.Component<PropsType, {}> {
  public render(): JSX.Element {
    const {bookmarkActions, bookmarkStore, bookmarkView} = this.props;

    return (
      <div>
        {
          bookmarkStore.bookmarks.map((bookmark: BookmarkModel) => (
            <Bookmark
              bookmark={bookmark}
              bookmarkActions={bookmarkActions}
              bookmarkView={bookmarkView}
              key={bookmark.url}
            />
          ))
        }
      </div>
    );
  }
}
